### ISSUE TITLE
regarding re-train the VGG19 model

### ISSUE THREAD

**Issue by surfreta** [gabrieldemarmiesse/heatmaps#3]
Hi,

Excuse me, I have to open a new thread regarding using your code posted in this [thread](https://github.com/gabrieldemarmiesse/heatmaps/issues/1).  Thanks.

Right now, I have 200 pairs of images for training. Each training image is of black and white, i.e., one channel only.  The mask image for training pair is also of size `256*256`, each of its pixel corresponds to one pixel in the original image.

In the code of setting the training data set,

```
X = np.random.uniform(0,256, (2000, 3, 256, 256))
X = preprocess_input(X) # VGG preprocessing
Y = np.around(np.random.uniform(0,1, (2000, 1, 64, 64)))

```
I have three questions, 

1)	does `VGG19 `model accept input with one channel only?
2)	In the code, Y was setup of having size of `64*64` due to two pooling layer. However, the mask image is of size` 256*256`. How can I map these 256*256 mask image to `64*64`?
3)	Here you specify to freeze the first two blocks of the VGG19 by 
```
for i in range(7):
    model.layers[i].trainable = False

```
Just curious, why we have to freeze the first two blocks instead of the first block? Are there any specific reason underlying this? Or generally if we only want to re-tune the model weights, it is a normal practice to change the top layer only. Is this understanding right?

Thank you very much.

**Comment by gabrieldemarmiesse** [comment]
Hi, 
Don't hesitate to ask if one answer is unclear :
1) short answer: vgg accept only RBG images so 3 channels. Long answer: you could try some black magic (like transform the one channel image into a RGB image but only with shades of grey, or changing the first layer of convolution of the vgg to conver the kernel to adapt to the formula of RGB->grey) but I strongly advise against because the vgg is used to use colors to take decisions. Grey images will confuse it. 
2)  This is a very open question. Since you can't use the vgg, I'd recommend not using poolings at all. This way, you won't have issues. This is the simplest way and should work fine. 
3) Freezing blocks is a way of preventing overfitting. It removes parameters to learn from the model and make it less flexible.  So you should freeze more layers if you have little data, and freeze less if you have lots of data. It also depends on the closeness to you problem with the original problem. For example if you had to find if a picture is the one of a dog or a cat, you'd freeze a lot of layers. If you have to find what's the type of lightning used in a picture, you'd need to freeze very few layers since imagenet competition isn't about that at all. Training one layer is the good default when the two problems

**Comment by gabrieldemarmiesse** [comment]
Sorry bad manip.

**Comment by gabrieldemarmiesse** [comment]
When the two problems are close. 
I hope I could be of some help. 
If I might add one last thing, keep your network small (32 neurons per layer max) and use and abuse of data augmentation. This is because you seem to have a small dataset. Overfitting will likely be an issue. 
Good luck, 
Gabriel.

**Comment by surfreta** [comment]
Gabriel, thank you so much for your detailed reply.

In accordance with your answer, it might not be a good idea for me to directly use VGG, since it relies on RGB image.  However, I can borrow the basic design of VGG, by removing the pooling layers. This way I can get output with the same size of mask image.  Since it is a new architecture, I generally have to re-train the model. Is my understanding correct?

Thanks,
surfreta

**Comment by gabrieldemarmiesse** [comment]
Yes. This is it. Don't forget to remove the flatten and dense layers too. It should look like this: 
InputLayer -> Conv2d - > Conv2d - >... ... ... - > Conv2d. With padding="same" at each convolution layer, you should have the input size equal to the output size. 

You're welcome. 
Gabriel
